hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1741 @@
+    /// Create a new [`Config`] in the default configuration
+    ///
+    /// This function will enable the corresponding APB clock, reset the
+    /// [`Sercom`] peripheral, and return a [`Config`] in the default
+    /// configuration:
+    ///
+    /// * [`EightBit`] [`CharSize`]
+    /// * No parity
+    /// * One stop bit
+    /// * LSB-first
+    ///
+    /// [`Config`] takes ownership of the [`Sercom`] and [`Pads`].
+    ///
+    /// Users must configure GCLK manually. The `freq` parameter represents the
+    /// GCLK frequency for this [`Sercom`] instance.
+    #[inline]
+    pub fn new(clk: &Clock, mut sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        sercom.enable_apb_clock(clk);
+        Self::default(sercom, pads, freq).bit_order(BitOrder::LsbFirst)
+    }
+
+    /// Create a new [`Config`] in the default configuration
+    #[inline]
+    fn default(sercom: P::Sercom, pads: P, freq: impl Into<Hertz>) -> Self {
+        let mut registers = Registers::new(sercom);
+        registers.swrst();
+
+        // Enable internal clock mode
+        registers.configure_mode();
+        registers.configure_pads(P::RXPO as u8, P::TXPO as u8);
+        registers.configure_charsize(EightBit::BITS);
+
+        Self {
+            registers,
+            pads,
+            chsize: PhantomData,
+            freq: freq.into(),
+        }
+    }
+}
+
+impl<P, C> Config<P, C>
+where
+    P: ValidPads,
+    C: CharSize,
+{
+    /// Change the [`Config`] [`CharSize`]
+    #[inline]
+    fn change<C2>(self) -> Config<P, C2>
+    where
+        C2: CharSize,
+    {
+        Config {
+            registers: self.registers,
+            pads: self.pads,
+            chsize: PhantomData,
+            freq: self.freq,
+        }
+    }
+
+    /// Trigger the [`Sercom`]'s SWRST and return a [`Config`] in the
+    /// default configuration.
+    #[inline]
+    pub fn reset(self) -> Config<P> {
+        Config::default(self.registers.free(), self.pads, self.freq)
+    }
+
+    /// Consume the [`Config`], reset the peripheral, and return the [`Sercom`]
+    /// and [`Pads`]
+    #[inline]
+    pub fn free(mut self) -> (P::Sercom, P) {
+        self.registers.swrst();
+        (self.registers.free(), self.pads)
+    }
+
+    /// Change the [`CharSize`].
+    #[inline]
+    pub fn char_size<C2: FixedCharSize>(mut self) -> Config<P, C2> {
+        self.registers.configure_charsize(C2::BITS);
+        self.change()
+    }
+
+    /// Change the [`CharSize`] to [`DynCharSize`]. The UART's character
+    /// size will be changed to the provided `C2`, which is a [`FixedCharSize`]
+    /// type, and can then be changed dynamically on an enabled [`Uart`]
+    /// without changing the underlying [`Config`]'s type through the
+    /// [`reconfigure`](Uart::reconfigure) method.
+    #[inline]
+    pub fn dyn_char_size<C2: FixedCharSize>(mut self) -> Config<P, DynCharSize> {
+        self.registers.configure_charsize(C2::BITS);
+        self.change()
+    }
```

Comment:
This seems weird to me. You're setting it to `DynCharSize`, but you still want a compile-time known type to use for the actual value? What if I don't know the value until run-time? Shouldn't there be a `CharSizeEnum` or something?

I let users set `DynLength` like any other `Length` value, with the `.length::<L2>()` method. But then I added `get_dyn_length` and `set_dyn_length` methods that are only implemented when the `Config` is in `DynLength` mode. I think that approach might make more sense here.

The only difference is that `DynLength` can be set to anything from 1-255, so a `u8` works well to represent it at run-time. But a value-level version of `CharSize` should probably be an `enum`.

Reply:
You're right, that was weird. I changed the `set_dyn_char_size` method to take a `CharSizeEnum`, and added a `get_dyn_char_size` method.